
syntax = "proto3";

package mapreduce;

service MapReduce {
    
  // A simple RPC.
  //
  // Extracts files from given Path, sorts into buckets, and returns the output file names as they are saved.
  rpc Map (MapRequest) returns (stream OutputPath) {}

  // A simple RPC.
  //
  // Extracts files from given Path, does word count, and returns the output file names as they are saved. 
  rpc Reduce (ReduceRequest) returns (stream OutputPath) {}

  // Server shutdown RPC.
  //
  rpc Stop (StopRequest) returns (ShutDownResponse) {}
}

message MapRequest {
  string input_path = 1;    // path to input files 
  string output_path = 2;   // path to save outputs 
  int32 M = 3;              // number of buckets  
}

message ReduceRequest {
  string input_path = 1;    // path to input files (e.g. output of map task)
  string output_path = 2;   // path to save outputs
}

message OutputPath {
  string path = 1;          // path to output file
}

message StopRequest {
  bool shouldstop = 1;      // indicator that task is finished and server should close 
}

message ShutDownResponse {
  bool isshutdown = 1;      // indicator that server has closed
}

// Command line to generate pb2 and pb2_grpc files: 
// python -m grpc_tools.protoc --proto_path protos --python_out=. --grpc_python_out=. protos/mapreduce.proto